controller/common/common.go

```diff
@@ -34,6 +39,15 @@ var (
 	KeyFunc = cache.DeletionHandlingMetaNamespaceKeyFunc
 )
 
+type RuntimeContext struct {
```

Comment:
Let us explain on how this will be used. In other words, code comments should help.

Reply:
I felt `ControllerOptions` will be a better name.
FYI `context` is Golang refers to something else.

Reply:
I like `ControllerOptions` as well. I also feel it is line with the naming conventions used in controller-runtime and client-go.

Reply:
hmm I would like to differentiate `options` (which seems as static configuration used to run something) from what actually is stored there (actually objects to interacts with k8s cluster) so I proposed :
1. I renamed `Options` to `Configuration`
2. I renamed `RuntimeContext` to `ControllerContext`   - I am not sure if `Context` is much better than `Options` but it describes `dynamic` nature of what it holds. 

I also tried to wrap some simple behavior (Start/Stop) there.

I am open for discussion :)

Reply:
How about having it named to Controller. In other words this is a common Controller instance that can be embedded in specific controller instances such as DecoratorController, CompositeController, etc. In addition, I believe common Controller displays all characteristics of a singleton & hence we should rework the code accordingly.

```go
package common

type controller struct {
  Resources dynamicdiscovery.Resources
  DynClient   ...
  ...
}

var once sync.Once

// singleton instance
var instance *controller

func ControllerInstance(..) *controller {
  if instance != nil {
   return instance
  }
  once.Do(func(){
    instance = &controller{
       ...
       ...
    }
  })
  return instance
}
```

Reply:
hmm  actually I do not think it would be good as controller in k8s world is something which have a resource and logic for managing it, and this object originated as just wrapping up arguments for `metacontroller's` (and later - `parentController's` and `decoratorcontroller's`)  - currently we have :
- `metacontroller`- for managing metacontroller crd's
- `parentController` / `decoratorController` for managing user provided CRD's

the type which we introducing has nothing to do with managing anything, it is just runtime helper.

Semantically, I see :
 - `configuration` - as some type for holding/manipulating static data
 - `context` - was idea of the same, but holding dynamic helper objects

Reply:
Well.. we meet might need to think further. I am reluctant to use context that is passed around and also having it as a field in the struct. 

We can think this to be ControllerBase. I think I am confusing all by proposing different names. So I will leave it to finally.

Reply:
> Well.. we meet might need to think further. I am reluctant to use context that is passed around and also having it as a field in the struct.
> 

yes, this it not ideal, I was thinking of having `Factory` having that, currently `metacontroller` is creating controllers, therefore it must have it as a struct to pass it...I will try to think how it could be temporary improved

Reply:
for now I will remove  `ConftrollerContext` from being field in struct in `metacontrollers` - I just pass required paramaters to `newDecorator/ParentController` via args

Reply:
@AmitKumarDas @fpetkovski I took a bit of step back, as it seems that end state would need more serious refactor, For now I tried to address Amit concern of not having `context` as part of struct.

I believe before reaching end state, some intermediate steps need to occur (of course in separate PR's). Some ideas are :
1. thing about making `sync` methods in both metacontrollers lighter, so move logic of creating `parent`/`decorator` controllers to retrospective factories
2.  similar thing for `customize manager`
3. I personally need to dig into `controller-runtime` and `kubebuilder` to see how they handle configuration and common stuff